File: app/src/main/java/com/elementary/tasks/core/view_models/notes/BaseNotesViewModel.java
package com.elementary.tasks.core.view_models.notes;

import android.app.Application;

import com.elementary.tasks.core.controller.EventControlFactory;
import com.elementary.tasks.core.data.models.Note;
import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.utils.CalendarUtils;
import com.elementary.tasks.core.view_models.BaseDbViewModel;
import com.elementary.tasks.core.view_models.Commands;
import com.elementary.tasks.notes.work.DeleteNoteFilesAsync;
import com.elementary.tasks.reminder.work.DeleteFilesAsync;

import androidx.annotation.NonNull;

/**
 * Copyright 2018 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
abstract class BaseNotesViewModel extends BaseDbViewModel {

    BaseNotesViewModel(Application application) {
        super(application);
    }

    public void deleteNote(@NonNull Note note) {
        isInProgress.postValue(true);
        run(() -> {
            getAppDb().notesDao().delete(note);
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.DELETED);
            });
            new DeleteNoteFilesAsync(getApplication()).execute(note.getKey());
        });
    }

    public void saveNote(@NonNull Note note) {
        isInProgress.postValue(true);
        run(() -> {
            getAppDb().notesDao().insert(note);
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.SAVED);
            });
        });
    }

    public void deleteReminder(@NonNull Reminder reminder) {
        isInProgress.postValue(true);
        run(() -> {
            EventControlFactory.getController(reminder).stop();
            getAppDb().reminderDao().delete(reminder);
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.UPDATED);
            });
            CalendarUtils.deleteEvents(getApplication(), reminder.getUniqueId());
            new DeleteFilesAsync(getApplication()).execute(reminder.getUuId());
        });
    }
}

File: app/src/main/java/com/elementary/tasks/core/view_models/notes/BaseNotesViewModel.kt
package com.elementary.tasks.core.view_models.notes

import androidx.lifecycle.viewModelScope
import com.elementary.tasks.core.data.dao.NotesDao
import com.elementary.tasks.core.data.models.ImageFile
import com.elementary.tasks.core.data.models.NoteWithImages
import com.elementary.tasks.core.utils.work.WorkerLauncher
import com.elementary.tasks.core.utils.Constants
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.elementary.tasks.core.view_models.BaseProgressViewModel
import com.elementary.tasks.core.view_models.Commands
import com.elementary.tasks.core.view_models.DispatcherProvider
import com.elementary.tasks.notes.work.DeleteNoteBackupWorker
import com.elementary.tasks.notes.work.NoteSingleBackupWorker
import kotlinx.coroutines.launch
import timber.log.Timber

abstract class BaseNotesViewModel(
  dispatcherProvider: DispatcherProvider,
  protected val workerLauncher: WorkerLauncher,
  protected val notesDao: NotesDao
) : BaseProgressViewModel(dispatcherProvider) {

  fun deleteNote(noteWithImages: NoteWithImages) {
    val note = noteWithImages.note ?: return
    postInProgress(true)
    viewModelScope.launch(dispatcherProvider.default()) {
      notesDao.delete(note)
      for (image in noteWithImages.images) {
        notesDao.delete(image)
      }
      workerLauncher.startWork(DeleteNoteBackupWorker::class.java, Constants.INTENT_ID, note.key)
      postInProgress(false)
      postCommand(Commands.DELETED)
    }
  }

  fun saveNoteColor(note: NoteWithImages, color: Int) {
    val v = note.note ?: return
    note.note?.color = color
    postInProgress(true)
    viewModelScope.launch(dispatcherProvider.default()) {
      v.updatedAt = DateTimeManager.gmtDateTime
      notesDao.insert(v)
      workerLauncher.startWork(NoteSingleBackupWorker::class.java, Constants.INTENT_ID, v.key)
      postInProgress(false)
      postCommand(Commands.SAVED)
    }
  }

  fun saveNote(note: NoteWithImages) {
    val v = note.note ?: return
    postInProgress(true)
    viewModelScope.launch(dispatcherProvider.default()) {
      v.updatedAt = DateTimeManager.gmtDateTime
      saveImages(note.images, v.key)
      notesDao.insert(v)
      workerLauncher.startWork(NoteSingleBackupWorker::class.java, Constants.INTENT_ID, v.key)
      postInProgress(false)
      postCommand(Commands.SAVED)
    }
  }

  protected fun saveImages(list: List<ImageFile>, id: String) {
    val oldList = notesDao.getImages(id)
    Timber.d("saveImages: ${oldList.size}")
    for (image in oldList) {
      Timber.d("saveImages: delete -> ${image.id}, ${image.noteId}")
      notesDao.delete(image)
    }
    if (list.isNotEmpty()) {
      val newList = list.map {
        it.noteId = id
        it
      }
      Timber.d("saveImages: new list -> $newList")
      notesDao.insertAll(newList)
    }
  }
}
